test: Karpenter Integration Test Migration RFC (#2144)
kubernetes-sigs/karpenter: designs/karpenter-integration-testing.md

# Karpenter - Integration Testing RFC

## Summary

> Historically, the AWS Karpenter Provider served as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. The kubernetes-sigs/karpenter (Core) repository has contained testing that is more limited, and primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with KWOKNodeClass for node simulation to validate basic provisioning and deprovisioning functionality. To best support specific Karpenter provider implementations, and to reduce the toil of test duplication, this document identifies core common tests that should be migrated from the AWS Karpenter provider implementation into the Core Karpenter repository. The test migration will help provider implementers by reducing the pain points of having to be aware of all functionality changes in the Karpenter core, having to continually update individual tests within their repositories, and by specifying a common set of features and capabilities that are expected from all Karpenter provider implementation. This document serves as a living specification for Karpenter's integration testing strategy and orchestration mechanisms. As the project evolves, both the design requirements and this documentation will be updated to reflect current best practices and project needs.

Reviewer: ```suggestion
Historically, the AWS Karpenter Provider served as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. The kubernetes-sigs/karpenter (Core) repository has contained testing that is more limited, and primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with KWOKNodeClass for node simulation to validate basic provisioning and deprovisioning functionality. To best support specific Karpenter provider implementations and to reduce the toil of test duplication, this document identifies core common tests that should be migrated from the AWS Karpenter provider implementation into the Core Karpenter repository. 

The test migration will help provider implementers by reducing the pain points of having to be aware of all functionality changes in the Karpenter core, having to continually update individual tests within their repositories, and by specifying a common set of features and capabilities that are expected from all Karpenter provider implementations. This document serves as a living specification for Karpenter's integration testing strategy and orchestration mechanisms. As the project evolves, both the design requirements and this documentation will be updated to reflect current best practices and project needs.
```

nit

> Historically, the AWS Karpenter Provider served as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. The kubernetes-sigs/karpenter (Core) repository has contained testing that is more limited, and primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with KWOKNodeClass for node simulation to validate basic provisioning and deprovisioning functionality. To best support specific Karpenter provider implementations, and to reduce the toil of test duplication, this document identifies core common tests that should be migrated from the AWS Karpenter provider implementation into the Core Karpenter repository. The test migration will help provider implementers by reducing the pain points of having to be aware of all functionality changes in the Karpenter core, having to continually update individual tests within their repositories, and by specifying a common set of features and capabilities that are expected from all Karpenter provider implementation. This document serves as a living specification for Karpenter's integration testing strategy and orchestration mechanisms. As the project evolves, both the design requirements and this documentation will be updated to reflect current best practices and project needs.

Reviewer: ```suggestion
Historically, the AWS Karpenter Provider served as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. The kubernetes-sigs/karpenter (Core) repository has contained testing that is more limited, and primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with KWOKNodeClass for node simulation to validate basic provisioning and deprovisioning functionality. To best support specific Karpenter provider implementations, and to reduce the toil of test duplication, this document identifies core common tests that should be migrated from the AWS Karpenter provider implementation into the Core Karpenter repository. The test migration will help provider implementers by reducing the pain points of having to be aware of all functionality changes in the Karpenter core, having to continually update individual tests within their repositories, and by specifying a common set of features and capabilities that are expected from all Karpenter provider implementation. 
```

Do we actually expect ourselves to maintain this doc as we add more tests -- my gut tells me that this is quickly going to get out of date

Reviewer 2: You have referred to this as a living document though - what do we plan to continue to update?

Reviewer: +1, we should just drop this line -- it's not strictly necessary and design docs are notorious for going out of date once they are merged

The testing framework for the AWS Karpenter Provider has several types of tests:

* Unit Tests
    * Run in a mock Kubernetes environment
    * Part of the regular developer workflow

* Integration Tests
    * Create EKS clusters and interact with a live API server and AWS APIs 
    * Run on every commit
* Scale Tests
    * Will validate Karpenter's performance at large scales
    * Focus on horizontal scaling workloads and resource provisioned by Karpenter
* Soak Tests 
    * Designed to identify resource usage regressions
    * Will target memory leaks and long-running stability issues

The kubernetes-sigs/karpenter repository mainly utilizes unit tests to validate functionality. 

## Problem 

> The testing architecture in the Karpenter ecosystem faces two critical challenges in its testing framework: having a  set of tests for all core functionality, and having a standard set of tests for all providers. The `kubernetes-sigs/karpenter` repository lacks comprehensive integration tests to validate core functionality across code changes, making it difficult to ensure that fundamental Karpenter behaviors remain intact as the codebase evolves. Additionally, cloud provider implementations of Karpenter lack a standardized conformance testing framework, meaning there's no systematic way to verify that provider-specific implementations maintain compatibility with `kubernetes-sigs/karpenter's` core requirements and expectations. It's worth noting that historically the `kubernetes-sigs/karpenter` repository relayed tests running within the AWS Karpenter. This document defines the appropriate placement of different test types across repositories, establishes clear testing responsibilities for both `kubernetes-sigs/karpenter` and cloud provider implementations, and provides a conformance testing framework that cloud providers can use to validate their implementations against kubernetes-sigs/karpenter's requirements. By addressing these gaps, we can ensure better code quality, maintain consistency across implementations, and provide a clear path for new cloud providers to validate their Karpenter implementations. 

Reviewer: ```suggestion
The testing architecture in the Karpenter ecosystem faces two critical challenges in its testing framework: having a set of tests for all core functionality and having a standard set of tests for all providers. The `kubernetes-sigs/karpenter` repository lacks comprehensive integration tests to validate core functionality across code changes, making it difficult to ensure that fundamental Karpenter behaviors remain intact as the codebase evolves.

Additionally, cloud provider implementations of Karpenter lack a standardized conformance testing framework, meaning there's no systematic way to verify that provider-specific implementations maintain compatibility with `kubernetes-sigs/karpenter's` core requirements and expectations. It's worth noting that historically the `kubernetes-sigs/karpenter` repository relyed on tests running within the AWS Karpenter. 

This document defines the appropriate placement of different test types across repositories, establishes clear testing responsibilities for both `kubernetes-sigs/karpenter` and cloud provider implementations, and provides a conformance testing framework that cloud providers can use to validate their implementations against kubernetes-sigs/karpenter's requirements.
```

nit: Consider dropping and using paragraphs as you have different thoughts

#### Kubernetes-sigs/karpenter

* Environment: Kind clusters with KWOK Karpenter Controller for simulated node provisioning
* Test Coverage:
    * NodePool Configuration
    * Standalone NodeClaim Configuration
    * Provisioning
    * Scheduling
    * Disruption Management
        * Consolidation
        * Drift (including NodePool changes)
        * Expiration
        * Emptiness
        * Node Repair Termination
    * Termination
    * Storage
    * Scale

#### Karpenter Cloud Provider

*  Environment: Non-Kind Cluster with in-cluster Cloud Provider Karpenter Controller
* Test Coverage:
    * OS Management
    * NodeClass Configuration
    * Disruption Handling
        * Drift (NodeClass)
        * Node Repair Conditions
        * Forceful Disruption Operations

> As part of the initial functionality integration testing, the cloud-agnostic tests from the AWS Karpenter Provider repository will be migrated to core Karpenter repository. These tests will be contained in the Kubernetes-sigs repository and made available for use by any cloud provider. The migration will focus on provider-neutral test cases including core provisioning behaviors, scaling operations, configuration validation, and general controller functionality. This consolidation will establish a standardized testing framework while reducing duplicate code across different provider implementations. By centralizing these common test scenarios, we can ensure consistent validation of critical functionality across all cloud providers while maintaining clear documentation and portable test utilities such as:

Reviewer: ```suggestion
As part of the initial functional integration testing, the cloud-agnostic tests from the AWS Karpenter Provider repository will be migrated to the core Karpenter repository. These tests will be contained in the `kubernetes-sigs/karpenter` repository and made available for use by any cloud provider. The migration will focus on provider-neutral test cases including core provisioning behaviors, scaling operations, configuration validation, and general controller functionality. This consolidation will establish a standardized testing framework while reducing duplicate code across different provider implementations. By centralizing these common test scenarios, we can ensure consistent validation of critical functionality across all cloud providers while maintaining clear documentation and portable test utilities such as:
```

nit

# Karpenter - Integration Testing RFC

> Note: This document serves as a living specification for Karpenter's integration testing strategy and orchestration mechanisms. As the project evolves, both the design requirements and this documentation will be updated to reflect current best practices and project needs.

Reviewer 3: i think this note could just be incorporated into the summary.

## Summary

> The AWS Karpenter Provider currently serves as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. In the kubernetes-sigs/karpenter (Core) repository, testing is more limited, primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with KWOKNodeClass for node simulation to validate basic provisioning and deprovisioning functionality. Each Karpenter Provider would currently need to need to maintain a similar set of integration tests to validate the functionality of any updates to the core project, creating testing duplication across the community. This is particularly painful as each cloud provider would need to be aware of functionality changes and create/update their integration tests.

Reviewer 3: given the note about this being a living document, i don't think we should say things like "currently". imo, talking about the connection to the original AWS tests is good, but i think we should say this more as a historical context to help readers understand how we got here. also, i think if we are summarizing we should talk about the outcomes of this document, eg:

```suggestion
Historically, the AWS Karpenter Provider served as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. The kubernetes-sigs/karpenter (Core) repository has contained testing that is more limited, and primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with KWOKNodeClass for node simulation to validate basic provisioning and deprovisioning functionality. To best support specific Karpenter provider implementations, and to reduce the toil of test duplication, this document identifies core common tests that should be migrated from the AWS Karpenter provider implementation into the Core Karpenter repository. The test migration will help provider implementers by reducing the pain points of having to be aware of all functionality changes in the Karpenter core, having to continually update individual tests within their repositories, and by specifying a common set of features and capabilities that are expected from all Karpenter provider implementations.
```

The testing framework for the AWS Karpenter Provider has several types of tests:

* Unit Tests
    * Run in a mock Kubernetes environment
    * Part of the regular developer workflow

* Integration Tests
    * Create EKS clusters and interact with a live API server and AWS APIs 
    * Run on every commit
* Scale Tests
    * Will validate Karpenter's performance at large scales
    * Focus on horizontal scaling workloads and resource provisioned by Karpenter
* Soak Tests 
    * Designed to identify resource usage regressions
    * Will target memory leaks and long-running stability issues

The kubernetes-sigs/karpenter repository mainly utilizes unit tests to validate functionality. 

## Problem 

> The current testing architecture in the Karpenter ecosystem faces two critical challenges in its testing framework. The `kubernetes-sigs/karpenter` repository lacks comprehensive integration tests to validate core functionality across code changes, making it difficult to ensure that fundamental Karpenter behaviors remain intact as the codebase evolves. Additionally, cloud provider implementations of Karpenter lack a standardized conformance testing framework, meaning there's no systematic way to verify that provider-specific implementations maintain compatibility with `kubernetes-sigs/karpenter's` core requirements and expectations. It's worth noting that `kubernetes-sigs/karpenter` repository relayed tests running within the AWS Karpenter. This document aims to define the appropriate placement of different test types across repositories, establish clear testing responsibilities for both `kubernetes-sigs/karpenter` and cloud provider implementations, and create an conformance testing framework that cloud providers can use to validate their implementations against kubernetes-sigs/karpenter's requirements. By addressing these gaps, we can ensure better code quality, maintain consistency across implementations, and provide a clear path for new cloud providers to validate their Karpenter implementations. 

Reviewer 3: minor cleanup nits, i think we can make this statement more definitive. eg:

```suggestion
The testing architecture in the Karpenter ecosystem faces two critical challenges in its testing framework: having a  set of tests for all core functionality, and having a standard set of tests for all providers. The `kubernetes-sigs/karpenter` repository lacks comprehensive integration tests to validate core functionality across code changes, making it difficult to ensure that fundamental Karpenter behaviors remain intact as the codebase evolves. Additionally, cloud provider implementations of Karpenter lack a standardized conformance testing framework, meaning there's no systematic way to verify that provider-specific implementations maintain compatibility with `kubernetes-sigs/karpenter's` core requirements and expectations. It's worth noting that historically the `kubernetes-sigs/karpenter` repository relayed tests running within the AWS Karpenter. This document defines the appropriate placement of different test types across repositories, establishes clear testing responsibilities for both `kubernetes-sigs/karpenter` and cloud provider implementations, and provides a conformance testing framework that cloud providers can use to validate their implementations against kubernetes-sigs/karpenter's requirements. By addressing these gaps, we can ensure better code quality, maintain consistency across implementations, and provide a clear path for new cloud providers to validate their Karpenter implementations. 
```

## Solution

> `kubernetes-sigs/karpenter` serves as the core repository, housing the majority of core functionality validation, including CEL validation for NodePool and NodeClaim. This repository also provides base integration tests that cloud providers can consume and adapt by injecting their own NodeClass implementations to verify compatibility with core functionality. Any additional validation rules introduced by cloud providers must be tested within their respective provider repositories. The responsibility boundary will be drawn as such: 

Reviewer 3: minor cleanup nits

```suggestion
The `kubernetes-sigs/karpenter` repository will serve as the primary location for a common test suite. The repository will contain tests for the majority of core functionality validation, including CEL validation for NodePool and NodeClaim. This repository will also provide base integration tests that cloud providers can consume and adapt by injecting their own NodeClass implementations to verify compatibility with core functionality. Any additional validation rules introduced by cloud providers must be tested within their respective provider repositories. The responsibility boundary will be drawn as such: 
```

#### Kubernetes-sigs/karpenter

* Environment: Kind clusters with KWOK Karpenter Controller for simulated node provisioning
* Test Coverage:
    * NodePool Configuration
    * Standalone NodeClaim Configuration
    * Provisioning
    * Scheduling
    * Disruption Management
        * Consolidation
        * Drift (including NodePool changes)
        * Expiration
        * Emptiness
        * Node Repair Termination
    * Termination
    * Storage
    * Scale

#### Karpenter Cloud Provider

*  Environment: Non-Kind Cluster with in-cluster Cloud Provider Karpenter Controller
* Test Coverage:
    * OS Management
    * NodeClass Configuration
    * Disruption Handling
        * Drift (NodeClass)
        * Node Repair Conditions
        * Forceful Disruption Operations

> As part of the initial functionality integration testing, we plan to migrate the cloud-agnostic tests from the AWS Karpenter Provider to this repository. These tests will be housed in the Kubernetes-sigs repository and made available for use by any cloud provider. The migration will focus on provider-neutral test cases including core provisioning behaviors, scaling operations, configuration validation, and general controller functionality. This consolidation effort aims to establish a standardized testing framework while reducing duplicate code across different provider implementations. By centralizing these common test scenarios, we can ensure consistent validation of critical functionality across all cloud providers while maintaining clear documentation and portable test utilities:

Reviewer 3: minor cleanup

```suggestion
As part of the initial functionality integration testing, the cloud-agnostic tests from the AWS Karpenter Provider repository will be migrated to core Karpenter repository. These tests will be contained in the Kubernetes-sigs repository and made available for use by any cloud provider. The migration will focus on provider-neutral test cases including core provisioning behaviors, scaling operations, configuration validation, and general controller functionality. This consolidation will establish a standardized testing framework while reducing duplicate code across different provider implementations. By centralizing these common test scenarios, we can ensure consistent validation of critical functionality across all cloud providers while maintaining clear documentation and portable test utilities such as:
```

##### Chaos 

* High Level Description: Validate that karpenter does not have any runaway scaling during provisioning or consolidation
* Source: https://github.com/aws/karpenter-provider-aws/tree/main/test/suites/chaos
* Test Cases:
    * Should not produce a runaway scale-up when consolidation is enabled
    * Should not produce a runaway scale-up when emptiness is enabled 

##### Consolidation

* High Level Description: Validate the basic consolidation action works as expected
* Source: https://github.com/aws/karpenter-provider-aws/tree/main/test/suites/consolidation
* Test Cases: 
    * LastPodEventTime
        * should update lastPodEventTime when pods are scheduled and removed
        * should update lastPodEventTime when pods go terminal
    * Budgets
        * should respect budgets for empty delete consolidation
        * should respect budgets for non-empty delete consolidation
        * should respect budgets for non-empty replace consolidation
        * should not allow consolidation if the budget is fully blocking
        * should not allow consolidation if the budget is fully blocking during a scheduled time
        * should consolidate nodes (delete)

>             * if the nodes are on-demand nodes

Reviewer 3: i wonder about the on-demand behavior, what if a provider does not have these type of instances?

Author: I would be curios to learn more about the offerings that are being passed back the cloud provider. But It could be possible for use to have filtering our specific test cases by cloud provider to avoid running specific tests.

            * should error when a requirement uses In but has no values
            * should error when a requirement uses an unknown operator
            * should error when Gt is used with multiple integer values
            * should error when Lt is used with multiple integer values
            * should error when consolidateAfter is negative
            * should succeed when ConsolidationPolicy=WhenEmptyOrUnderutilized is used with consolidateAfter
            * should error when minValues for a requirement key is negative
            * should error when minValues for a requirement key is zero
            * should error when minValues for a requirement key is more than 50
            * should error when minValues for a requirement key is greater than the values specified within In operator

##### Nodeclaim

* High Level Description: Validate that action taken by the nodeclaim controllers. This validates standalone claims and the garbage collection mechanisms  
* Source: https://github.com/aws/karpenter-provider-aws/tree/main/test/suites/nodeclaim
* Test Cases: 
    * Standalone NodeClaim 

>         * should create a standard NodeClaim with the c instance family 

Reviewer 3: what is the "c instance family", is this an AWS specific reference?

Author: Yeah, this is a AWS specific reference. The idea here would that we can launch any generic instance type. I will drop this test since we will get this validation for free with the other tests

# Karpenter - Integration Testing RFC

## Summary

> Historically, the AWS Karpenter Provider served as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. The kubernetes-sigs/karpenter (Core) repository has contained testing that is more limited, and primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with KWOKNodeClass for node simulation to validate basic provisioning and deprovisioning functionality. To best support specific Karpenter provider implementations and to reduce the toil of test duplication, this document identifies core common tests that should be migrated from the AWS Karpenter provider implementation into the Core Karpenter repository. 

Reviewer 2: nits:
- To maintain clarity, I think we should refer to the repos by their GitHub paths rather than descriptions
- Wording suggestion for clarity: we have a lot of functional / unit tests in this repo, just not integration tests
```suggestion
Historically, the [`aws/karpenter-provider-aws`](https://github.com/aws/karpenter-provider-aws) repository has served as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. The [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter) repository has contained limited integration testing, primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with `KWOKNodeClass` for node simulation to validate basic provisioning and deprovisioning functionality. To best support specific Karpenter provider implementations and to reduce the toil of test duplication, this document identifies core common tests that should be migrated from [`aws/karpenter-provider-aws`](https://github.com/aws/karpenter-provider-aws) provider implementation into [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter). 
```

> The test migration will help provider implementers by reducing the pain points of having to be aware of all functionality changes in the Karpenter core, having to continually update individual tests within their repositories, and by specifying a common set of features and capabilities that are expected from all Karpenter provider implementations. This document serves as a living specification for Karpenter's integration testing strategy and orchestration mechanisms. As the project evolves, both the design requirements and this documentation will be updated to reflect current best practices and project needs.

Reviewer 2: nit
```suggestion
The test migration will help provider implementers by reducing the pain points of having to be aware of all functionality changes in [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter), having to continually update individual tests within their repositories, and by specifying a common set of features and capabilities that are expected from all Karpenter provider implementations. This document serves as a living specification for Karpenter's integration testing strategy and orchestration mechanisms. As the project evolves, both the design requirements and this documentation will be updated to reflect current best practices and project needs.
```

> The testing framework for the AWS Karpenter Provider has several types of tests:

Reviewer 2: ```suggestion
The testing framework for [`aws/karpenter-provider-aws`](https://github.com/aws/karpenter-provider-aws) has several types of tests:
```

> * Unit Tests
>     * Run in a mock Kubernetes environment
>     * Part of the regular developer workflow
> 
> * Integration Tests
>     * Create EKS clusters and interact with a live API server and AWS APIs 
>     * Run on every commit
> * Scale Tests
>     * Will validate Karpenter's performance at large scales
>     * Focus on horizontal scaling workloads and resource provisioned by Karpenter
> * Soak Tests 
>     * Designed to identify resource usage regressions
>     * Will target memory leaks and long-running stability issues

Reviewer 2: nit: consistent tense

Also, can you add the details for when scale and soak tests are ran for parity with unit and integration? I don't know the frequency off the top of my head.
```suggestion
* Unit Tests
    * Runs in a mock Kubernetes environment
    * Part of the regular developer workflow (`make test`)
    * Runs as a pre-merge CI check on GitHub

* Integration Tests
    * Creates EKS clusters and interact with a live API server and AWS APIs 
    * Runs against every commit merged to `main`
    
* Scale Tests
    * Validates Karpenter's performance at large scales
    * Focuses on horizontally scaled workloads and resources provisioned by Karpenter
    
* Soak Tests 
    * Identifies resource usage regressions, specifically memory leaks and long-running stability issues
```

> The kubernetes-sigs/karpenter repository mainly utilizes unit tests to validate functionality. 

Reviewer 2: ```suggestion
The [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter) repository mainly utilizes unit tests to validate functionality. 
```

## Problem 

> The testing architecture in the Karpenter ecosystem faces two critical challenges in its testing framework: having a set of tests for all core functionality and having a standard set of tests for all providers. The `kubernetes-sigs/karpenter` repository lacks comprehensive integration tests to validate core functionality across code changes, making it difficult to ensure that fundamental Karpenter behaviors remain intact as the codebase evolves.

Reviewer 2: ```suggestion
The testing architecture in the Karpenter ecosystem faces two critical challenges in its testing framework: having a set of tests for all core functionality and having a standard set of tests for all providers. The [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter) repository lacks comprehensive integration tests to validate core functionality across code changes, making it difficult to ensure that fundamental Karpenter behaviors remain intact as the codebase evolves.
```

> Additionally, cloud provider implementations of Karpenter lack a standardized conformance testing framework, meaning there's no systematic way to verify that provider-specific implementations maintain compatibility with `kubernetes-sigs/karpenter's` core requirements and expectations. It's worth noting that historically the `kubernetes-sigs/karpenter` repository relyed on tests running within the AWS Karpenter. 

Reviewer 2: ```suggestion
Additionally, cloud provider implementations of Karpenter lack a standardized conformance testing framework, meaning there's no systematic way to verify that provider-specific implementations maintain compatibility with [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter)'s core requirements and expectations. It's worth noting that historically the [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter) repository relied on tests running within [`aws/karpenter-provider-aws`](https://github.com/aws/karpenter-provider-aws). 
```

> This document defines the appropriate placement of different test types across repositories, establishes clear testing responsibilities for both `kubernetes-sigs/karpenter` and cloud provider implementations, and provides a conformance testing framework that cloud providers can use to validate their implementations against kubernetes-sigs/karpenter's requirements.

Reviewer 2: ```suggestion
This document defines the appropriate placement of different test types across repositories, establishes clear testing responsibilities for both [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter) and cloud provider implementations, and provides a conformance testing framework that cloud providers can use to validate their implementations against [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter)'s requirements.
```

## Solution

> The `kubernetes-sigs/karpenter` repository will serve as the primary location for a common test suite. The repository will contain tests for the majority of core functionality validation, including CEL validation for NodePool and NodeClaim. This repository will also provide base integration tests that cloud providers can consume and adapt by injecting their own NodeClass implementations to verify compatibility with core functionality. Any additional validation rules introduced by cloud providers must be tested within their respective provider repositories. The responsibility boundary will be drawn as such: 

Reviewer 2: ```suggestion
The [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter) repository will serve as the primary location for a common test suite. The repository will contain tests for the majority of core functionality validation, including CEL validation for NodePool and NodeClaim. This repository will also provide base integration tests that cloud providers can consume and adapt by injecting their own NodeClass implementations to verify compatibility with core functionality. Any additional validation rules introduced by cloud providers must be tested within their respective provider repositories. The responsibility boundary will be drawn as such: 
```

> #### Kubernetes-sigs/karpenter

Reviewer 2: ```suggestion
#### kubernetes-sigs/karpenter
```

* Environment: Kind clusters with KWOK Karpenter Controller for simulated node provisioning
* Test Coverage:
    * NodePool Configuration
    * Standalone NodeClaim Configuration
    * Provisioning
    * Scheduling
    * Disruption Management
        * Consolidation
        * Drift (including NodePool changes)
        * Expiration
        * Emptiness
        * Node Repair Termination
    * Termination
    * Storage
    * Scale

#### Karpenter Cloud Provider

*  Environment: Non-Kind Cluster with in-cluster Cloud Provider Karpenter Controller
* Test Coverage:
    * OS Management
    * NodeClass Configuration
    * Disruption Handling
        * Drift (NodeClass)
        * Node Repair Conditions
        * Forceful Disruption Operations

> As part of the initial functional integration testing, the cloud-agnostic tests from the AWS Karpenter Provider repository will be migrated to core Karpenter repository. These tests will be contained in the `kubernetes-sigs/karpenter` repository and made available for use by any cloud provider. The migration will focus on provider-neutral test cases including core provisioning behaviors, scaling operations, configuration validation, and general controller functionality. This consolidation will establish a standardized testing framework while reducing duplicate code across different provider implementations. By centralizing these common test scenarios, we can ensure consistent validation of critical functionality across all cloud providers while maintaining clear documentation and portable test utilities such as:

Reviewer 2: ```suggestion
As part of the initial functional integration testing, the cloud-agnostic tests from [`aws/karpenter-provider-aws`](https://github.com/aws/karpenter-provider-aws) will be migrated to [`kubernetes-sigs/karpenter`](https://github.com/kubernetes-sigs/karpenter), where they will be made available for use by any cloud provider. The migration will focus on provider-neutral test cases including core provisioning behaviors, scaling operations, configuration validation, and general controller functionality. This consolidation will establish a standardized testing framework while reducing duplicate code across different provider implementations. By centralizing these common test scenarios, we can ensure consistent validation of critical functionality across all cloud providers while maintaining clear documentation and portable test utilities such as:
```

# Karpenter - Integration Testing RFC

> Note: This document serves as a living specification for Karpenter's integration testing strategy and orchestration mechanisms. As the project evolves, both the design requirements and this documentation will be updated to reflect current best practices and project needs.

Reviewer 2: Will this document be expanded in the future to detail how a cloudprovider can use the upstream integration tests in their own environment with their own NodeClass?

Author: No, we will not be including that process in this document. We will have a living README document in the test suite that will outline how cloud provider and utilize these tests

## Summary

> The AWS Karpenter Provider currently serves as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. In the kubernetes-sigs/karpenter (Core) repository, testing is more limited, primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with KWOKNodeClass for node simulation to validate basic provisioning and deprovisioning functionality. Each Karpenter Providers would currently need to need to maintain a set smiler set of integration tests to validate the functionality of any updates to the core project, creating testing duplication across the community. This is particularly painful as each cloud provider would need to be aware of functionality changes and create/update their integration tests 

Reviewer 2: spelling
```suggestion
The AWS Karpenter Provider currently serves as the primary testing ground for Karpenter's core functionality, with comprehensive integration tests covering provisioning, scheduling, disruption handling, networking configurations, and AWS-specific feature implementations. In the kubernetes-sigs/karpenter (Core) repository, testing is more limited, primarily utilizing the KWOK (Kubernetes Without Kubelet) Provider with KWOKNodeClass for node simulation to validate basic provisioning and deprovisioning functionality. Each Karpenter Provider would currently need to need to maintain a similar set of integration tests to validate the functionality of any updates to the core project, creating testing duplication across the community. This is particularly painful as each cloud provider would need to be aware of functionality changes and create/update their integration tests.
```

The testing framework for the AWS Karpenter Provider has several types of tests:

* Unit Tests
    * Run in a mock Kubernetes environment
    * Part of the regular developer workflow

* Integration Tests

>     * Create real clusters and interact with actual APIs

Reviewer 2: I'd emphasize that these are the cloudprovider APIs (e.g. AWS). Both the unit and integration tests interact with a real API server and real k8s apis.

>     * Validate the cluster changes 

Reviewer 2: What does this mean?

    * Run on every commit

> * Scale Tests

Reviewer 4: Is there a plan on what should happen to the existing performance test suite in the `kubernetes-sigs/karpenter` repo?

Author: We have some planned work to understand performance of Karpenter. However, that will be out of the scope of this RFC

    * Will validate Karpenter's performance at large scales

>     * Focus on horizontal scaling workloads and resource management

Reviewer 2: What resources are we talking about here? Resources provisioned by Karpenter as NodeClaims or resources consumed by the Karpenter controller pod?

Author: This would be resources provisioned by Karpenter. Updated to reflect that

* Soak Tests 
    * Designed to identify resource usage regressions
    * Will target memory leaks and long-running stability issues

The kubernetes-sigs/karpenter repository mainly utilizes unit tests to validate functionality. 

## Problem 

> The current testing architecture in the Karpenter ecosystem faces two critical challenges in its testing framework. The `kubernetes-sigs/karpenter` repository lacks comprehensive integration tests to validate core functionality across code changes, making it difficult to ensure that fundamental Karpenter behaviors remain intact as the codebase evolves. Additionally, cloud provider implementations of Karpenter lack a standardized conformance testing framework, meaning there's no systematic way to verify that provider-specific implementations maintain compatibility with `kubernetes-sigs/karpenter's` core requirements and expectations. This document aims to define the appropriate placement of different test types across repositories, establish clear testing responsibilities for both `kubernetes-sigs/karpenter` and cloud provider implementations, and create an conformance testing framework that cloud providers can use to validate their implementations against kubernetes-sigs/karpenter's requirements. By addressing these gaps, we can ensure better code quality, maintain consistency across implementations, and provide a clear path for new cloud providers to validate their Karpenter implementations. 

Reviewer 2: > The `kubernetes-sigs/karpenter` repository lacks comprehensive integration tests to validate core functionality across code changes, making it difficult to ensure that fundamental Karpenter behaviors remain intact as the codebase evolves.

I'd emphasize that this is only true without relying on external cloudprovider integration tests. It's not as though we've been flying blind this entire time - it's just that integration tests remained in the AWS provider as an artifact from splitting Karpenter into a core library and separate provider implementations. I think the main value here is that it allows user to validate changes in a cloudprovider-agnostic environment.

Author: That's a good point, I'll add that in

## Solution

> The Kubernetes-sigs/karpenter serves as the core repository, housing the majority of core functionality validation, including CEL validation for NodePool and NodeClaim. This repository also provides base integration tests that cloud providers can consume and adapt by injecting their own NodeClass implementations to verify compatibility with core functionality. Any additional validation rules introduced by cloud providers must be tested within their respective provider repositories. The responsibility boundary will be drawn as such: 

Reviewer 4: ```suggestion
`kubernetes-sigs/karpenter` serves as the core repository, housing the majority of core functionality validation, including CEL validation for NodePool and NodeClaim. This repository also provides base integration tests that cloud providers can consume and adapt by injecting their own NodeClass implementations to verify compatibility with core functionality. Any additional validation rules introduced by cloud providers must be tested within their respective provider repositories. The responsibility boundary will be drawn as such: 
```

    * NodeClass Configuration
    * Disruption Handling
        * Drift (NodeClass)
        * Node Repair Conditions
        * Forceful Disruption Operations

As part of the initial functionality integration testing, we plan to migrate the cloud-agnostic tests from the AWS Karpenter Provider to this repository. These tests will be housed in the Kubernetes-sigs repository and made available for use by any cloud provider. The migration will focus on provider-neutral test cases including core provisioning behaviors, scaling operations, configuration validation, and general controller functionality. This consolidation effort aims to establish a standardized testing framework while reducing duplicate code across different provider implementations. By centralizing these common test scenarios, we can ensure consistent validation of critical functionality across all cloud providers while maintaining clear documentation and portable test utilities:

##### Chaos 

* High Level Description: Validate that karpenter does not have any runaway scaling during provisioning or consolidation
* Source: https://github.com/aws/karpenter-provider-aws/tree/main/test/suites/chaos
* Test Cases:

>     * Should not produce a runaway  scale-up when consolidation is enabled

Reviewer 4: ```suggestion
    * Should not produce a runaway scale-up when consolidation is enabled
```

* Source: https://github.com/aws/karpenter-provider-aws/tree/main/test/suites/nodeclaim
* Test Cases: 
    * Standalone NodeClaim 
        * should create a standard NodeClaim with the c instance family 
        * should create a standard NodeClaim based on resource requests
        * should create a NodeClaim propagating all NodeClaim spec details
        * should remove the cloudProvider NodeClaim when the cluster NodeClaim is deleted
        * should delete a NodeClaim from the node termination finalizer 
        * should create a NodeClaim with custom labels passed through the userData
        * should delete a NodeClaim after the registration timeout when the node doesn’t register
        * should delete a NodeClaim if it references a NodeClass that doesn’t exist
        * should delete a NodeClaim if it references a NodeClass that isn’t Ready 
    * Garbage Collection 
        * should succeed to garbage collect an Instance that was launched by a NodeClaim but has no Instance mapping 
        * should succeed to garbage collect an Instance that was deleted without cluster’s knowledge 

##### Scale

> * High Level Description: Scale test to see any regression with Karpenter performance. These are high level tests that only really validate the provisioning and speed of getting pods to a ready state.  

Reviewer 4: Does it make sense to port over these scale tests? If these tests are to validate the speed of getting pods to a ready state, wouldn't that be a cloud-provider specific problem since I'm assuming different infrastructure have different performance benchmarks?

Or is the expectation that all cloud-providers should be able to provision nodes (I think I see `expectedNodes=500` for one of the tests) in a similar/reasonable amount of time?

Author: We are planning on doing some benchmarking work on our scheduler, so I'm not sure if it makes sense to move these tests from the AWS provider. @DerekFrank is planning on doing that work
